src/allmydata/node.py

```diff
@@ -428,6 +434,74 @@ def get_config_path(self, *args):
             os.path.join(self._basedir, *args)
         )
 
+    def get_introducer_configuration(self):
+        """
+        Get configuration for introducers.
+
+        :return {unicode: (unicode, FilePath)}: A mapping from introducer
+            petname to a tuple of the introducer's fURL and local cache path.
+        """
+        introducers_yaml_filename = self.get_private_path("introducers.yaml")
+        introducers_filepath = FilePath(introducers_yaml_filename)
+
+        def get_cache_filepath(petname):
+            return FilePath(
+                self.get_private_path("introducer_{}_cache.yaml".format(petname)),
+            )
+
+        try:
+            with introducers_filepath.open() as f:
+                introducers_yaml = safe_load(f)
+                if introducers_yaml is None:
+                    raise EnvironmentError(
+                        EPERM,
+                        "Can't read '{}'".format(introducers_yaml_filename),
+                        introducers_yaml_filename,
+                    )
+                introducers = {
+                    petname: config["furl"]
+                    for petname, config
+                    in introducers_yaml.get("introducers", {}).items()
+                }
+                assert all(isinstance(k, str) for k in introducers.keys())
```

Comment:
For sure.  I think Tahoe-LAFS is the only place I ever write asserts ... because it is loaded with them already.

If we all think they're not a good idea maybe we should make it official, forbid new ones, and start grinding down the old ones (either directly or opportunistically with other efforts)?

Reply:
Not commenting on other uses, but asserts seem fine for truly asserting "a thing that shouldn't happen" but we definitely shouldn't write new code where the asserts are validating user-provided data. (Maybe "non-str keys" are really "impossible" from yaml here, but ..?)

Reply:
Yea I put them in because I didn't know what was coming back from the yaml library, esp on Py 2 vs Py 3, and I wanted it to fail as early as possible.

But if we think something is really impossible and is also worth proving hasn't happened, then don't we want that proof to always hold, not just when someone leaves asserts on?

Reply:
Nah, asserts are fine for that i'd say .. I just misread it at first to be verification-type code